rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/model/ParametricSummary.java

```diff
@@ -0,0 +1,500 @@
+        super(config, exec, columns);
+        this.contrib = contrib;
+    }
+
+    public ParametricSummary createSummary(ISourceLocation file, Versioned<ITree> tree, Position cursor) {
+        return new OndemandSummary(file, tree, cursor);
+    }
+
+    public <T> @Nullable InterruptibleFuture<List<T>> createSummaryThenLookup(
+            ISourceLocation file, Versioned<ITree> tree, Position cursor,
+            SummaryLookup<T> lookup) {
+
+        return lookup.apply(new OndemandSummary(file, tree, cursor), cursor);
+    }
+
+    public class OndemandSummary implements ParametricSummary {
+        private final ISourceLocation file;
+        private final Versioned<ITree> tree;
+        private final Position cursor;
+
+        public OndemandSummary(ISourceLocation file, Versioned<ITree> tree, Position cursor) {
+            this.file = file;
+            this.tree = tree;
+            this.cursor = cursor;
+        }
+
+        @Override
+        @SuppressWarnings("deprecation") // For `MarkedString`
+        public @Nullable InterruptibleFuture<List<Either<String, MarkedString>>> getDocumentation(Position cursor) {
+            return get(config.providesDocumentation, cursor, contrib::documentation, ParametricSummaryFactory::mapValueToString, DOCUMENTATION);
+        }
+
+        @Override
+        public @Nullable InterruptibleFuture<List<Location>> getDefinitions(Position cursor) {
+            return get(config.providesDefinitions, cursor, contrib::definitions, columns::mapValueToLocation, DEFINITIONS);
+        }
+
+        @Override
+        public @Nullable InterruptibleFuture<List<Location>> getReferences(Position cursor) {
+            return get(config.providesReferences, cursor, contrib::references, columns::mapValueToLocation, REFERENCES);
+        }
+
+        @Override
+        public @Nullable InterruptibleFuture<List<Location>> getImplementations(Position cursor) {
+            return get(config.providesImplementations, cursor, contrib::implementations, columns::mapValueToLocation, IMPLEMENTATIONS);
+        }
+
+        @Override
+        public InterruptibleFuture<List<Diagnostic>> getMessages() {
+            return InterruptibleFuture.completedFuture(Collections.emptyList());
+        }
+
+        @Override
+        public void invalidate() {
+            // Nothing to invalidate
+        }
+
+        private <T> @Nullable InterruptibleFuture<List<T>> get(boolean provides, Position cursor,
+                OndemandCalculator calculator, Function<IValue, T> valueMapper, String logName) {
+
+            // Note on second disjunct: To ensure that this summary can't be
+            // misused if it's accidentally leaked elsewhere, this summary
+            // provides information only if the cursor is *identical* (not just
+            // equal) to the cursor when this summary was created.
+            if (!provides || this.cursor != cursor) {
```

Comment:
great! maybe we should log this case? so that we say: hey unexpected use happens, help future us in debugging.

Reply:
Added